### PR TITLE
add llvm load store

### PR DESCRIPTION
**Pull request by edg-l**

### CODE REVIEW

File: melior/src/dialect/llvm.rs [hunk 1]
```diff
@@ -101,6 +104,102 @@ pub fn undef<'c>(result_type: Type<'c>, location: Location<'c>) -> Operation<'c>
...
+    pub noalias_scopes: Option<ArrayAttribute<'c>>,
+    pub tbaa: Option<ArrayAttribute<'c>>,
+}
+
+impl<'c> LoadStoreOptions<'c> {
+    pub fn fill_attributes(
+        &self,
+        context: &'c Context,
+        attributes: &mut Vec<(Identifier<'c>, Attribute<'c>)>,
+    ) {
```
Comment (raviqqe) [inline melior/src/dialect/llvm.rs]: How about this?

```suggestion
    fn into_attributes(
        self,
        context: &'c Context,
    ) -> Vec<(Identifier<'c>, Attribute<'c>)> {
```

File: melior/src/dialect/llvm.rs [hunk 2]
```diff
@@ -101,6 +104,102 @@ pub fn undef<'c>(result_type: Type<'c>, location: Location<'c>) -> Operation<'c>
...
+    let mut attributes = vec![];
+
+    if let Some(extra_options) = extra_options {
+        extra_options.fill_attributes(context, &mut attributes);
+    }
+
+    OperationBuilder::new("llvm.store", location)
+        .add_operands(&[value, addr])
+        .add_attributes(&attributes)
+        .build()
```
Comment (raviqqe) [inline melior/src/dialect/llvm.rs]:
```suggestion
    extra_options: LoadStoreOptions<'c>,
) -> Operation<'c> {
    OperationBuilder::new("llvm.store", location)
        .add_operands(&[value, addr])
        .add_attributes(&extra_options.into_attributes())
        .build()
```

File: melior/src/dialect/llvm.rs [hunk 3]
```diff
@@ -101,6 +104,102 @@ pub fn undef<'c>(result_type: Type<'c>, location: Location<'c>) -> Operation<'c>
...
+
+    if let Some(extra_options) = extra_options {
+        extra_options.fill_attributes(context, &mut attributes);
+    }
+
+    OperationBuilder::new("llvm.load", location)
+        .add_operands(&[addr])
+        .add_attributes(&attributes)
+        .add_results(&[r#type])
+        .build()
```
Comment (raviqqe) [inline melior/src/dialect/llvm.rs]:
```suggestion
    extra_options: LoadStoreOptions<'c>,
) -> Operation<'c> {
    OperationBuilder::new("llvm.load", location)
        .add_operands(&[addr])
        .add_attributes(&extra_options.into_attributes())
        .add_results(&[r#type])
        .build()
```

**Review by raviqqe — COMMENTED** [review]
Thank you for the feature additions again! 😄

File: melior/src/dialect/llvm.rs [hunk 4]
```diff
@@ -101,6 +104,102 @@ pub fn undef<'c>(result_type: Type<'c>, location: Location<'c>) -> Operation<'c>
         .build()
 }
 
+#[derive(Debug, Default)]
+pub struct LoadStoreOptions<'c> {
```
Comment (raviqqe) [inline melior/src/dialect/llvm.rs:108]: This is out of scope of this PR but we should probably use the builder pattern instead. Have you seen this pattern of options somewhere else?

**Comment by edg-l** [comment]
I applied your suggestions, but I don't know how exactly you want the builder pattern so I'll leave it up to you on another pr if thats right?

**Comment by edg-l** [comment]
(sorry forgot to fix the tests according to the changes)

**Pull request merged**